### PR TITLE
shadow.set_password: handle case where user has no shadow entry

### PR DESCRIPTION
**Pull request by terminalmage**
Before this fix, this function would not set the password, because its
method of updating the shadow file is to read all the lines into a list,
iterate over that list and update the line for the desired user, and
rewrite the list of lines to /etc/shadow. The problem with this method
is that in the corner cases where the shadow entry doesn't exist (don't
ask), the password is not set, since there is no line for that user.

This fixes that corner case by appending a shadow entry when no line
exists for the specified user.

### CODE REVIEW

**Review by waynew — DISMISSED** [review]
Hey Erik, thanks for the PR!

I'm sure there's a fun backstory there 🙃 

The fix looks great! Just a couple of minor tweaks to the tests would be awesome.

---

I checked out the test failures and it looks like they're infrastructure issues, so I'm expecting they should be resolved on the next test run.

File: tests/unit/modules/test_shadow.py [hunk 1]
```diff
@@ -64,6 +67,81 @@ def test_gen_password(self):
...
+        lines = filehandles[1].writelines_calls[0]
+        # Should only have the same two users in the file
+        assert len(lines) == 2
+        # The first line should be unchanged
+        assert lines[0] == 'foo:orighash:17955::::::\n'
+        # The second line should have the new password hash
+        assert lines[1].split(':')[:2] == [user, password]
+
+        #
+        # CASE 2: Corner case: no /etc/shadow entry for user
```
[the remark this answers is not in the document]
Comment (terminalmage) [reply tests/unit/modules/test_shadow.py]: I realize that's probably _your_ preference, but with there historically being issues with test results being stored in memory for the life of the test run, leading to frequent issues with test VMs running out of memory, I don't agree.

File: salt/modules/shadow.py [hunk 2]
```diff
@@ -379,21 +386,31 @@ def set_password(name, password, use_usermod=False, root=None):
         if not os.path.isfile(s_file):
             return ret
         lines = []
+        user_found = False
+        lstchg = six.text_type(
```
[the remark this answers is not in the document]
Comment (terminalmage) [reply salt/modules/shadow.py]: I don't think so. If you're developing for a *NIX platform and do not understand what the UNIX epoch is...

File: tests/unit/modules/test_shadow.py [hunk 3]
```diff
@@ -64,6 +67,81 @@ def test_gen_password(self):
...
+        data = {
+            '/etc/shadow': salt.utils.stringutils.to_bytes(textwrap.dedent('''\
+                foo:orighash:17955::::::
+                bar:somehash:17955::::::
+                ''')),
+            '*': Exception('Attempted to open something other than /etc/shadow'),
+        }
+        isfile_mock = MagicMock(
+            side_effect=lambda x: True if x == '/etc/shadow' else DEFAULT)
+        shadow_info_mock = MagicMock(return_value={'passwd': 'newhash'})
```
[the remark this answers is not in the document]
Comment (terminalmage) [reply tests/unit/modules/test_shadow.py]: Fair enough.

https://github.com/saltstack/salt/pull/56044/commits/e2afef76f7a01eae1f19a1299ab0d37d2fedcaef

File: salt/modules/shadow.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (waynew) [reply salt/modules/shadow.py]: 🤷‍♂️ while I agree with that sentiment, I also agree that explicit is better than implicit. I always encourage the latter. Especially since future me is usual the one maintaining the software I wrote and he always appreciates it 🤣

File: tests/unit/modules/test_shadow.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sagetherage) [reply tests/unit/modules/test_shadow.py]: I am nearly certain we have addressed the test VMs running out of memory, and let me get an expert to weigh-in.

[the remark this answers is not in the document]
Comment (dwoz) [reply tests/unit/modules/test_shadow.py]: @terminalmage for what it's worth, those issues have been addressed.

**Review by dwoz — DISMISSED** [review]

File: tests/unit/modules/test_shadow.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (terminalmage) [reply tests/unit/modules/test_shadow.py]: @dwoz From this very PR: https://jenkinsci.saltstack.com/job/pr-centos6-py2/job/PR-56044/2/testReport/junit/unit.grains.test_core/CoreGrainsTestCase/test_virtual_has_virtual_grain/

**Comment by terminalmage** [comment]
Is there anything still holding this up/

**Review by dwoz — APPROVED** [review]
